pkg/dag/packages.go

```diff
@@ -0,0 +1,222 @@
+				return err
+			}
+			c := &Configuration{
+				Configuration: buildc,
+				Path:          p,
+			}
+
+			name := c.Package.Name
+			if name == "" {
+				log.Fatalf("no package name in %q", path)
+			}
+			if err := pkgs.addConfiguration(name, c); err != nil {
+				return err
+			}
+			pkgs.addPackage(name)
+
+			for _, prov := range c.Package.Dependencies.Provides {
+				name, version := packageNameFromProvides(prov)
+				if err := pkgs.addNameVersionDescription(name, version, fmt.Sprintf("PROVIDED BY %s", c.Package.Name), c.Path); err != nil {
+					return err
+				}
+			}
+
+			for i := range c.Subpackages {
+				subpkg := c.Subpackages[i]
+				if subpkg.Name == "" {
+					log.Fatalf("empty subpackage name at index %d for package %q", i, c.Package.Name)
+				}
+				if err := pkgs.addConfiguration(subpkg.Name, c); err != nil {
+					return err
+				}
+
+				for _, prov := range subpkg.Dependencies.Provides {
+					name, version := packageNameFromProvides(prov)
+					if err := pkgs.addNameVersionDescription(name, version, fmt.Sprintf("PROVIDED BY %s", c.Package.Name), c.Path); err != nil {
+						return err
+					}
+				}
+
+				// TODO: resolve deps via `uses` for subpackage pipelines.
+			}
+			// Resolve all `uses` used by the pipeline. This updates the set of
+			// .environment.contents.packages so the next block can include those as build deps.
+			pctx := &build.PipelineContext{
+				Context: &build.Context{
+					Configuration: *c.Configuration,
+				},
+				Package: &c.Package,
+			}
+			for i := range c.Pipeline {
+				s := c.Pipeline[i]
+				if err := s.ApplyNeeds(pctx); err != nil {
+					return fmt.Errorf("unable to resolve needs for package %s: %w", name, err)
+				}
+				c.Environment.Contents.Packages = pctx.Context.Configuration.Environment.Contents.Packages
+			}
+		}
+
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return pkgs, nil
+}
+
+// Config returns the Melange configuration for the package, provides or subpackage with the given name,
+// if the package is present in the Graph. If it's not present, Config returns
+// nil. Providing the name of a subpackage will return the configuration of the
+// subpackage's origin package.
+func (p Packages) Config(name string) []*Configuration {
+	if p.configs == nil {
+		// this would be unexpected
+		return nil
+	}
+
+	if c, ok := p.configs[name]; ok {
+		list := make([]*Configuration, 0, len(c))
+		list = append(list, c...)
+		return list
+	}
+
+	return nil
+}
+
+// PkgInfo returns the buildp.Package struct
+func (p Packages) PkgInfo(pkgName string) (*build.Package, error) {
+	config := p.Config(pkgName)
+	if len(config) == 0 {
+		log.Println("no config for package:", pkgName)
+		return nil, nil
+	}
+	c := config[0]
+	if pkgName != c.Package.Name {
+		return nil, nil
+	}
+	return &c.Package, nil
+}
+
+// Packages returns a slice of the names of all packages, sorted alphabetically.
+func (p Packages) Packages() []string {
+	allPackages := make([]string, 0)
```

Comment:
nit: We can set the capacity to `len(p.packages)` here if the goal is to avoid allocations

Reply:
Done